---
title: "Babybiome analysis"
output: html_document
date: '2023-05-10'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```


This Rmarkdown describes the TRACS analysis of the mother and baby metagenomic sequencing data originally published in Shao et al., 2019 and Shao et al., 2024

> Shao, Y. et al. Stunted microbiota and opportunistic pathogen colonization in caesarean-section birth. Nature 574, 117â€“121 (2019).

> Shao, Y. et al. Primary succession of Bifidobacteria drives pathogen resistance in neonatal microbiota assembly. Nat. Microbiol. (2024).
  

## Running TRACS

The TRACS alignment was performed on each pair of FASTQ files using the GTDB rs207 reference database (Parks et al., 2022) with the following commands. The reads were initially cleaned and trimmed as described in the original papers.

The following Bash commands accurately describe the execution process for running TRACS. However, they are not configured to run directly from this Rmarkdown document. Instead, these commands were executed separately on a server.

The TRACS align algorithm was then run on each pair of read files as

```
tracs align -i -i .${out_dir}_1.fastq.gz ${out_dir}_2.fastq.gz -p $out_dir -o $out_dir -t 15 --database ./gtdb-rs207.genomic-reps.dna.k51.sbt.zip --refseqs ./GTDB/gtdb_genomes_reps_r207/ --keep-all
```

The resulting alignments were then combined into individual species Multiple Sequence Alignments using the TRACS combine command

```
mkdir combined_alignments
tracs combine -i * -o combined_alignments/ -t 20
```

Finally, SNP distances were calculated using the distance command

```
for f in *_combined.fasta.gz
do
prefix=$(basename $f _combined.fasta.gz)
tracs distance --msa $f -o "${prefix}_distance.csv" --snp_threshold 5000 --filter -t 30
done
```

## Load results and generate plots

Note: some functions and code use the previous name for the TRACS algorithm (Trac'm).

```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(data.table)
library(ggthemes)
library(ggbeeswarm)
library(patchwork)
source("./calculate_snp_cutoff.R")
```

Load metadata and results

```{r}
meta <- fread("data/babybiome/metadata_BabyBiome1_2_2387.csv") %>% as_tibble()

sourmash_hits <- fread("./data/babybiome/combined_metadata.csv.gz") %>% as_tibble()

dist_files <- Sys.glob("./data/babybiome/*distance.csv.gz")
distances <- map_dfr(dist_files, ~{
  fread(.x, colClasses = c('character', 'character', 'integer', 'integer','integer','integer','integer','integer')) %>% 
    as_tibble() %>%
    add_column(species=gsub("_distance.*", "", gsub(".*babybiome/", "", .x)), .before=1)
})

tax <- rbind(fread("./data/bac120_metadata_r207.tsv.gz") %>% as_tibble(),
             fread("./data/ar53_metadata_r207.tsv.gz") %>% as_tibble())
  
tax <- tax %>%
  add_column(strain=gsub(".*s__", "", tax$gtdb_taxonomy)) %>%
  add_column(accession2=gsub("^[A-Z]*_", "", tax$accession), .before = 1)

tax$phylum <- gsub("_.*", "", gsub(";.*", "", gsub(".*p__", "", tax$gtdb_taxonomy)))

distances$strain <- tax$strain[match(distances$species, tax$accession2)]
```

Add relationships between pairs and merge metadata with distances

```{r}
meta$subject <- ifelse(meta$Time_point=="Mother", paste0(meta$Subject_ID, "_mother"), meta$Subject_ID)

distances$subjectA <- meta$subject[match(distances$sampleA, meta$Lane_ID)]
distances$subjectB <- meta$subject[match(distances$sampleB, meta$Lane_ID)]
distances$famA  <- meta$Family_ID[match(distances$sampleA, meta$Lane_ID)]
distances$famB  <- meta$Family_ID[match(distances$sampleB, meta$Lane_ID)]
distances$hospA  <- meta$Hospital[match(distances$sampleA, meta$Lane_ID)]
distances$hospB  <- meta$Hospital[match(distances$sampleB, meta$Lane_ID)]

distances$relationship <- map2_chr(distances$sampleA, distances$sampleB, ~{
  indexA <- which(meta$Lane_ID==.x)
  indexB <- which(meta$Lane_ID==.y)
  
  if (meta$subject[indexA]==meta$subject[indexB]) {
    if ("Mother" %in% c(meta$Time_point[indexA], meta$Time_point[indexB])){
      return("Same person (Mother)")  
    } else {
      return("Same person (Infant)")
    }
  } else if (meta$Family_ID[indexA]==meta$Family_ID[indexB]) {
    if ("Mother" %in% c(meta$Time_point[indexA], meta$Time_point[indexB])){
      return(paste0("Same family (Mother/Infant): ", meta$Delivery[indexA]))  
    } else {
      return("Same family (Infant/Infant)")
    }
  } else if (meta$Hospital[indexA]==meta$Hospital[indexB]){
    return("Same hospital") }
  else {
    return("Different hospital")  
  }
})

distances$time_diff <- NA
meta$time <- as.numeric(meta$Time_point)
distances$time_diff[distances$relationship=="Same person (Infant)"] <- abs(meta$time[match(distances$sampleA, meta$Lane_ID)] -
                                                                 meta$time[match(distances$sampleB, meta$Lane_ID)])[distances$relationship=="Same person (Infant)"]
```

Estimate species specific SNP distances using the mixture distribution approach. Samples taken from the same infant are used as the 'close' dataset while samples from separate hospitals are used as the 'distant' dataset.

```{r}
sourmash_hits$subject <- meta$subject[match(sourmash_hits$sample, meta$Lane_ID)]
sourmash_hits$time <- meta$Time_point[match(sourmash_hits$sample, meta$Lane_ID)]
sourmash_hits$delivery <- meta$Delivery[match(sourmash_hits$sample, meta$Lane_ID)]
sourmash_hits$Hospital <- meta$Hospital[match(sourmash_hits$sample, meta$Lane_ID)]

species_counts <- sourmash_hits %>% filter(!is.na(frac_N)) %>%
  group_by(accession, species, subject, Hospital) %>%
  summarise(
    count=n(),
  ) %>% 
  arrange(-count)

within_pairwise_counts <- species_counts %>% 
  group_by(accession, species) %>%
  summarise(
    pw_counts = sum(count*(count-1)/2)
  ) %>%
  arrange(-pw_counts)

species_hosp_counts <- species_counts %>%
  group_by(accession, species, Hospital) %>%
  summarise(count=sum(count))

between_hospital_pairwise_counts <- species_hosp_counts %>% 
  group_by(accession, species) %>%
  summarise(
    pw_counts = (sum(count)*(sum(count)-1)/2) - sum(count*(count-1)/2)
  ) %>%
  arrange(-pw_counts)

within_hosp_counts <- species_counts %>%
  group_by(accession, species, Hospital) %>%
  summarise(
    t_counts = (sum(count)*(sum(count)-1)/2) - sum(count*(count-1)/2)
  ) %>%
  group_by(accession, species) %>%
   summarise(
    pw_counts = sum(t_counts)
  ) %>%
  arrange(-pw_counts)


distances$pid <- 1-distances$`filtered SNP distance`/distances$`sites considered`
```

```{r}
tb <- sort(table(distances$species))

combined <- mixture_snp_cutoff(distances$`filtered SNP distance`[distances$relationship=='Same person (Infant)'], 
                       distances$`filtered SNP distance`[distances$relationship=='Different hospital'], 
                       upper.tail = 1-1e-3)

thresholds <- map_dfr(unique(distances$species), ~{
  
  if (sum(distances$species==.x)<30){
    return(
      tibble(
        species=.x,
        snp_threshold=NA,
        lambda=NA,
        k=NA,
        estimated_fp=NA,
        youden_cutoff=NA,
        quantile_cutoff=NA
      )
    )
  }
  
  within <- distances %>%
    filter(species==.x) %>% 
    filter(relationship=='Same person (Infant)')
  
  between <- distances %>%
    filter(species==.x) %>% 
    filter(relationship=='Different hospital')
  
  res <- mixture_snp_cutoff(within$`filtered SNP distance`, 
                       between$`filtered SNP distance`, 
                       upper.tail = 1-1e-3) %>%
    add_column(species=.x, .before = 1)
  
  youden <- youden_snp_cutoff(within$`filtered SNP distance`, 
                       between$`filtered SNP distance`)
  res$youden_cutoff <- youden$snp_threshold
  
  if (nrow(within)>5){
    res$quantile_cutoff <- floor(min(quantile(within$`filtered SNP distance`, 0.95),
    quantile(between$`filtered SNP distance`, 0.05)))
  } else {
    res$quantile_cutoff <- NA
  }
  
  return(res)

})

thresholds$snp_threshold[is.na(thresholds$lambda)] <- combined$snp_threshold
thresholds$snp_threshold <- thresholds$snp_threshold*3

thresholds$species_name <- tax$strain[match(thresholds$species, tax$accession2)]

distances$snp_threshold <- thresholds$snp_threshold[match(distances$species, thresholds$species)]

pdfthresh <- thresholds %>% 
  filter(!is.na(lambda))
keep <- pdfthresh$species[order(-abs(pdfthresh$snp_threshold-pdfthresh$youden_cutoff))][1:3]


pdfthresh <- pdfthresh %>%
  pivot_longer(cols = c("snp_threshold", "youden_cutoff"))
colnames(pdfthresh)[[6]] <- 'strain'

pdfthreshdist <- distances %>% 
  filter(species %in% keep) %>%
  filter(relationship %in% c('Same person (Infant)', 'Different hospital'))
pdfthresh <- pdfthresh %>% filter(species %in% keep)
```

Count close transmission for each relationship type

```{r}
distances_filt_thresh <- distances %>%
  filter(`filtered SNP distance` <= snp_threshold)
meta_filt <- meta[meta$Lane_ID %in% c(distances$sampleA, distances$sampleB),]

m <- combn(meta_filt$Lane_ID, 2)
count_distance_filt <- tibble(sampleA = m[1,], sampleB = m[2,])
count_distance_filt$relationship <- map2_chr(count_distance_filt$sampleA, count_distance_filt$sampleB, ~{
  indexA <- which(meta$Lane_ID==.x)
  indexB <- which(meta$Lane_ID==.y)
  
  if (meta$subject[indexA]==meta$subject[indexB]) {
    if ("Mother" %in% c(meta$Time_point[indexA], meta$Time_point[indexB])){
      return("Same person (Mother)")  
    } else {
      return("Same person (Infant)")
    }
  } else if (meta$Family_ID[indexA]==meta$Family_ID[indexB]) {
    if ("Mother" %in% c(meta$Time_point[indexA], meta$Time_point[indexB])){
      return(paste0("Same family (Mother/Infant): ", meta$Delivery[indexA]))  
    } else {
      return("Same family (Infant/Infant)")
    }
  } else if (meta$Hospital[indexA]==meta$Hospital[indexB]){
    return("Same hospital") }
  else {
    return("Different hospital")  
  }
})

count_distance_filt$pair <- map2_chr(count_distance_filt$sampleA, 
                                     count_distance_filt$sampleB, 
                                     ~ paste(sort(c(.x, .y)), collapse = '-'))

distances_filt_thresh$pair <- map2_chr(distances_filt_thresh$sampleA, 
                                       distances_filt_thresh$sampleB, 
                                       ~ paste(sort(c(.x, .y)), collapse = '-'))

nonzero_count_distances_filt <- distances_filt_thresh %>% 
  # filter(!same_study) %>%
  group_by(pair) %>%
  summarise(
    n=sum(!is.na(`filtered SNP distance`)),
    relationship=unique(relationship)
  )

count_distance_filt$n <- nonzero_count_distances_filt$n[match(count_distance_filt$pair, nonzero_count_distances_filt$pair)]
count_distance_filt$n[is.na(count_distance_filt$n)] <- 0

levs <- (count_distance_filt %>%
  group_by(relationship) %>%
  summarise(mean=mean(n)) %>%
  arrange(mean))$relationship

pdf <- count_distance_filt %>% filter(relationship!="Same person (Infant)")
pdf$relationship <- factor(pdf$relationship, levels = levs)

ggplot(pdf, aes(x=n, y=..density.., fill=relationship)) +
  geom_histogram(binwidth = 1, position = 'dodge') +
  scale_y_sqrt() +
  scale_fill_brewer(type = 'qual', palette = 3) +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(), legend.background = element_blank(), legend.position = 'none') +
  xlab("Number of recently transmitted strains") +
  ylab("Fraction of pairwise relatioships")

# ggsave("figures/babybiome_shared_strain_counts.png", width = 10, height = 7)
# ggsave("figures/babybiome_shared_strain_counts.pdf", width = 10, height = 7)
```

we now estimate colonisation rates by species for mother/infant pairs

```{r}
distances_filt_thresh$delivery_A <- meta$Delivery[match(distances_filt_thresh$sampleA, meta$Lane_ID)]

donor_pair_count <- meta_filt %>% 
  # filter(Delivery=='Vaginal') %>%
  group_by(Family_ID, Delivery) %>%
  summarise(
    mother_id=subject[Time_point=="Mother"],
    is_mother=any(Time_point=="Mother"),
    n_children=length(unique(subject))-1
  ) %>%
  filter(is_mother)

donor_pair_count$sampleID <- meta$Lane_ID[match(donor_pair_count$mother_id, meta$subject)]

count_of_trans_opportunities <- sourmash_hits %>%
  filter(time=="Mother") %>%
  group_by(accession, delivery) %>%
  summarise(
    count = n(),
    possible_transmissions=sum(donor_pair_count$n_children[donor_pair_count$sampleID %in% sample])
  )

species_pair_counts <- map2_dfr(donor_pair_count$sampleID, donor_pair_count$n_children, ~{
  tibble(
    species=(sourmash_hits %>% filter(sample==.x) %>% filter(!is.na(frac_N)))$accession,
    count=.y
  )
}) %>%
  group_by(species) %>%
  summarise(
    total=sum(count)
  )


donor_recipient_dists <- distances_filt_thresh %>%
  filter(relationship %in% c("Same family (Mother/Infant): Vaginal", "Same family (Mother/Infant): Caesarean"))


count_distances_filt_mother_child <- donor_recipient_dists %>%
  filter(`filtered SNP distance`<snp_threshold) %>%
  group_by(species, strain, delivery_A) %>%
  summarise(
    n=length(unique(famA))
  )

count_distances_filt_mother_child$total <- count_of_trans_opportunities$possible_transmissions[match(
  paste(count_distances_filt_mother_child$species, count_distances_filt_mother_child$delivery_A),
  paste(count_of_trans_opportunities$accession, count_of_trans_opportunities$delivery))]
```

Generate plot

```{r}
pdf <- count_distances_filt_mother_child

## Maximum likelihood approach
pdf$ml_rate <- map2_dbl(pdf$n, pdf$total, ~ binom.test(x = .x, n = .y, conf.level = 0.95)$estimate)
pdf$ml_lower <- map2_dbl(pdf$n, pdf$total, ~ binom.test(x = .x, n = .y, conf.level = 0.95)$conf.int[[1]])
pdf$ml_upper <- map2_dbl(pdf$n, pdf$total, ~ binom.test(x = .x, n = .y, conf.level = 0.95)$conf.int[[2]])

keep <- (pdf %>%
  group_by(strain) %>%
  summarise(
    ttotal=sum(total)
  ) %>%
  filter(ttotal>=20))$strain

pdf <- pdf %>% filter(strain %in% keep)
pdf$strain <- factor(pdf$strain, levels = rev(unique(pdf$strain[order(-pdf$ml_rate)])))

cols <- setNames(c('#377eb8','#4daf4a','#984ea3','#ff7f00'),
                 c("Firmicutes", "Proteobacteria", "Actinobacteriota", "Bacteroidota"))

pdf$phylum <- tax$phylum[match(pdf$species, tax$accession2)]

# delivery_cols <- c('#3C5488FF', '#4DBBD5FF')
delivery_cols <- c('#b2df8a', '#33a02c')

gg1 <- ggplot(pdf, aes(x=ml_rate, y=strain, col=delivery_A, group=delivery_A)) +
  geom_point(size=2, position=ggstance::position_dodgev(height=0.9)) +
  geom_errorbarh(aes(xmin=ml_lower, xmax=ml_upper), position=ggstance::position_dodgev(height=0.9)) +
  # facet_wrap(~delivery_A, nrow = 1) +
  scale_y_discrete(
    breaks = pdf$strain,
    labels = setNames(pdf$strain, pdf$phylum)
  ) +
  scale_color_manual(values = delivery_cols) +
  scale_x_continuous(limits = c(0,0.4)) +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(), legend.position = 'none') +
  theme(axis.text.y = element_text(
    color = cols[pdf$phylum[!duplicated(pdf$strain)]])) +
  xlab("Transmission rate") +
  ylab("Species (GTDB)")

count_of_trans_opportunities$strain <- tax$strain[match(count_of_trans_opportunities$accession, tax$accession2)]
  
pdf2 <- count_of_trans_opportunities %>%
  filter(strain %in% pdf$strain)
pdf2$strain <- factor(pdf2$strain, levels = levels(pdf$strain))
  
gg2 <- ggplot(pdf2, aes(x=possible_transmissions, y=strain, fill=delivery)) +
  geom_col() +
  scale_fill_manual(values = delivery_cols) +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.y = element_blank()
        ) +
  xlab("Total pairs")


gg1 + gg2 + patchwork::plot_layout(nrow = 1)

# ggsave("figures/babybiome_mother_child_colonisation_rates_vaginal_delivery_mode.png", width = 10, height = 9)
# ggsave("figures/babybiome_mother_child_colonisation_rates_vaginal_delivery_mode.pdf", width = 10, height = 9)
```

To investigate persistence across different time points, we conducted a Chi-squared test to determine if the persistence of the strain significantly differs between time point 7 and the subsequent time points 21 and late infancy. Persistence rates are estimated assuming a Binomial distribution.

```{r}
estimate_rates <- map_dfr(c(7,21,'late_infancy'), function(time2) {
  donor_pair_count <- meta_filt %>% 
    group_by(Family_ID, subject) %>%
    summarise(
      timpointA_id=ifelse(sum(Time_point=='4')>0, Lane_ID[Time_point=='4'], NA),
      timpointB_id=ifelse(sum(Time_point==time2)>0, Lane_ID[Time_point==time2], NA)
    ) %>%
    filter(!is.na(timpointA_id) & !is.na(timpointB_id))
  
  species_pair_counts <- map_dfr(donor_pair_count$timpointA_id, ~{
    tibble(
      species=(sourmash_hits %>% filter(sample==.x) %>% filter(!is.na(frac_N)))$accession,
      count=1
    )
  }) %>%
    group_by(species) %>%
    summarise(
      total=sum(count)
    )
  
  distances_filt_thresh$is_either_time <- map2_lgl(distances_filt_thresh$sampleA, distances_filt_thresh$sampleB, ~{
    any(c(meta$Time_point[meta$Lane_ID==.x], meta$Time_point[meta$Lane_ID==.y])=='4') &
      any(c(meta$Time_point[meta$Lane_ID==.x], meta$Time_point[meta$Lane_ID==.y])==time2)
  })
  
  
  donor_recipient_dists <- distances_filt_thresh %>%
    filter(relationship=="Same person (Infant)") %>%
    filter(is_either_time)
  
  count_distances_filt_same_child <- donor_recipient_dists %>%
    filter(`filtered SNP distance`<snp_threshold) %>%
    group_by(species, strain) %>%
    summarise(
      n=length(unique(subjectA))
    )
  
  count_distances_filt_same_child$total <- species_pair_counts$total[match(count_distances_filt_same_child$species,
                                                                           species_pair_counts$species)]
  
  ## Maximum likelihood approach
  count_distances_filt_same_child$ml_rate <- map2_dbl(count_distances_filt_same_child$n, count_distances_filt_same_child$total, ~ binom.test(x = .x, n = .y, conf.level = 0.95)$estimate)
  count_distances_filt_same_child$ml_lower <- map2_dbl(count_distances_filt_same_child$n, count_distances_filt_same_child$total, ~ binom.test(x = .x, n = .y, conf.level = 0.95)$conf.int[[1]])
  count_distances_filt_same_child$ml_upper <- map2_dbl(count_distances_filt_same_child$n, count_distances_filt_same_child$total, ~ binom.test(x = .x, n = .y, conf.level = 0.95)$conf.int[[2]])
  
  
  count_distances_filt_same_child$strain <- factor(count_distances_filt_same_child$strain, levels = count_distances_filt_same_child$strain[order(count_distances_filt_same_child$ml_rate)])
  
  pdf <- count_distances_filt_same_child %>% filter(total>=10)

  return(pdf %>% add_column(`Time period` = paste0("4-", time2)))
})

estimate_rates$`Time period` <- factor(estimate_rates$`Time period`, levels = c("4-7", "4-21", "4-late_infancy"))

sig <- bind_rows(estimate_rates %>%
  group_by(species) %>%
  group_map(~{
    start <- .x %>% filter(`Time period`=='4-7')
    mid <- .x %>% filter(`Time period`=='4-21')
    end <- .x %>% filter(`Time period`=='4-late_infancy')
    
    if (nrow(start)<=0){
      return(tibble())
    }
    
    df <- tibble()
    
    if (nrow(mid)>0){
      cs_mid <- chisq.test(matrix(c(start$n, start$total-start$n, mid$n, mid$total-mid$n),
                       byrow = TRUE, nrow = 2, dimnames = list(time=c('start', 'end'), 
                                                               persisted=c('TRUE','FALSE'))))
      df <- rbind(df, tibble(strain=unique(.x$strain),
                             p.value=cs_mid$p.value,
                             time='4-21'))
    }
    
    if (nrow(end)>0){
      cs_end <- chisq.test(matrix(c(start$n, start$total-start$n, end$n, end$total-end$n),
                       byrow = TRUE, nrow = 2, dimnames = list(time=c('start', 'end'), 
                                                               persisted=c('TRUE','FALSE'))))
      df <- rbind(df, tibble(strain=unique(.x$strain),
                             p.value=cs_end$p.value,
                             time='4-late_infancy'))
    }
    
    
    return(df)
    
  }))

estimate_rates$issig <- paste(estimate_rates$strain, estimate_rates$`Time period`) %in% paste(sig$strain, sig$time)[sig$p.value<0.05]

estimate_rates$phylum <- tax$phylum[match(estimate_rates$species, tax$accession2)]

estimate_rates <- estimate_rates %>% arrange(-as.numeric(strain))

cols <- setNames(c('#377eb8','#4daf4a','#984ea3','#ff7f00'),
                 c("Firmicutes", "Proteobacteria", "Actinobacteriota", "Bacteroidota"))

tmp <- rowSums(table(estimate_rates$species, estimate_rates$`Time period`))

filt_estimate_rates <- estimate_rates %>% 
  filter(species %in% names(tmp[tmp>1])) %>%
  filter(total>=10)

ggplot(filt_estimate_rates, 
       aes(x=ml_rate, y=strain, col=issig)) +
  geom_point(size=2) +
  geom_errorbarh(aes(xmin=ml_lower, xmax=ml_upper)) +
  facet_wrap(~`Time period`, nrow = 1) +
  scale_y_discrete(
    breaks = filt_estimate_rates$strain,
    labels = setNames(filt_estimate_rates$strain, filt_estimate_rates$phylum)
  ) +
  scale_color_manual(values = c('black', 'red')) +
  theme_clean(base_size = 16, base_family = "Arial") +
  theme(plot.background = element_blank(),
        legend.position = 'none') +
  theme(axis.text.y = element_text(color = cols[filt_estimate_rates$phylum[!duplicated(filt_estimate_rates$strain)]])) +
  xlab("Estimated rate of persistance between timepoints") +
  ylab("Species (GTDB)")

# ggsave("figures/strain_persistence.png", width = 12, height = 8)
# ggsave("figures/strain_persistence.pdf", width = 12, height = 8, device = cairo_pdf)
```

### Comparison with StrainPhlAn

We now compare with the transmission estimates originally inferred using StrainPhlAn as part of the Babybiome 2 project (Shao et al., 2024). A full description of the how StrainPhlAn was run is given in that publication.

```{r}
prev_bb2_meta <- fread("./data/StrainPhlAn4_BBS12_result_dir/metadata_spa.tsv") %>% 
  as_tibble()
prev_bb2_species <- fread("./data/StrainPhlAn4_BBS12_result_dir/mpa_vOct22_CHOCOPhlAnSGB_202212_species.txt.gz", col.names = c('SGB', 'tax')) %>% 
  as_tibble()

sphlan_trans_thresholds <- map_dfr(Sys.glob("./data/StrainPhlAn4_BBS12_result_dir/species_transmission_events_dir/*.info.gz"), ~ {
  lines <- read_lines(.x)
  tibble(
    SGB=gsub("\\.info.*","", gsub(".*t__", "", .x)),
    threshold = as.numeric(gsub("Selected strain-transmission threshold: ", "",
                    lines[grepl("Selected strain-transmission threshold", lines)][[1]]))
  )
})

sphlan_distances <- map_dfr(Sys.glob("./data/StrainPhlAn4_BBS12_result_dir/spa_nGD_outdir/*nGD.tsv.gz"), ~{
  SGB <- gsub("_nGD\\.tsv.*","", gsub(".*t__", "", .x))
  if (!SGB %in% sphlan_trans_thresholds$SGB) return(tibble())
  sgb_threshold <- sphlan_trans_thresholds$threshold[sphlan_trans_thresholds$SGB==SGB]
  read_tsv(.x, col_names = c('sampleA', 'sampleB', 'distance')) %>% 
    # filter(distance<=sgb_threshold) %>%
    add_column(.before = 1, SGB=SGB) %>%
    add_column(sgb_threshold=sgb_threshold) %>%
    dplyr::mutate(
      sampleA_ordered = pmin(sampleA, sampleB),
      sampleB_ordered = pmax(sampleA, sampleB)
    ) %>%
    dplyr::select(-sampleA, -sampleB) %>%
    dplyr::rename(sampleA = sampleA_ordered, sampleB = sampleB_ordered)
})

meta$Lane_ID_temp <- gsub("_clean", "", meta$Lane_ID)

sphlan_distances$relationship <- map2_chr(sphlan_distances$sampleA, sphlan_distances$sampleB, ~{
  indexA <- which(meta$Lane_ID_temp==.x)
  indexB <- which(meta$Lane_ID_temp==.y)
  
  if (meta$subject[indexA]==meta$subject[indexB]) {
    if ("Mother" %in% c(meta$Time_point[indexA], meta$Time_point[indexB])){
      return("Same person (Mother)")  
    } else {
      return("Same person (Infant)")
    }
  } else if (meta$Family_ID[indexA]==meta$Family_ID[indexB]) {
    if ("Mother" %in% c(meta$Time_point[indexA], meta$Time_point[indexB])){
      return(paste0("Same family (Mother/Infant): ", meta$Delivery[indexA]))  
    } else {
      return("Same family (Infant/Infant)")
    }
  } else if (meta$Hospital[indexA]==meta$Hospital[indexB]){
    return("Same hospital") }
  else {
    return("Different hospital")  
  }
})

sphlan_distances$pair <-paste(sphlan_distances$sampleA, sphlan_distances$sampleB, sep = '-')
sphlan_distances_filtered <- sphlan_distances %>% filter(distance <= sgb_threshold)

sphlan_count_distance <- sphlan_distances %>% 
  group_by(pair) %>%
  summarise(
    n=n(),
    relationship=unique(relationship)
  )

sphlan_nonzero_count_distances_filt <- sphlan_distances_filtered %>% 
  group_by(pair) %>%
  summarise(
    n=n(),
    relationship=unique(relationship)
  )

sphlan_count_distance$count <- sphlan_nonzero_count_distances_filt$n[match(sphlan_count_distance$pair, 
                                                                                sphlan_nonzero_count_distances_filt$pair)]

sphlan_count_distance$count[is.na(sphlan_count_distance$count)] <- 0

sphlan_count_distance$relationship <- factor(sphlan_count_distance$relationship, 
                                                  levels = names(sort(table(sphlan_count_distance$relationship), decreasing = TRUE)))


pdf <- rbind(distances_filt_thresh %>% 
  group_by(relationship) %>%
  summarise(
    count=n()
  ) %>% add_column(method='Tracm', .before = 1),
  sphlan_distances_filtered %>% 
  group_by(relationship) %>%
  summarise(
    count=n()
  ) %>% add_column(method='StrainPhlan', .before = 1))

pdf$relationship <- factor(pdf$relationship, unique(pdf$relationship[order(pdf$count)]))
ggplot(pdf, aes(x=relationship, y=count, fill=method)) +
  geom_col(position = 'dodge') +
  coord_flip() +
  scale_y_sqrt() +
  scale_fill_brewer(palette = 6, type = 'qual') +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank()) +
  xlab("Relationship") +
  ylab("Number of inferred transmissions")


# ggsave("./figures/bbiome_sphlan_vs_tracm_total_counts.png", width = 12, height = 10)
# ggsave("./figures/bbiome_sphlan_vs_tracm_total_counts.pdf", width = 12, height = 10)

sgb_to_gtdb <- read_tsv("./data/ianiro/mpa_vOct22_CHOCOPhlAnSGB_202212_SGB2GTDB.tsv.gz", col_names = c("SGB", "tax"))

sphlan_distances_filtered$tax <- sgb_to_gtdb$tax[match(sphlan_distances_filtered$SGB, sgb_to_gtdb$SGB)]
sphlan_distances_filtered$strain <- tax$strain[match(sphlan_distances_filtered$tax, tax$gtdb_taxonomy)]

sphlan_species_tran_counts <- sphlan_distances_filtered %>% 
  group_by(relationship, strain) %>%
  summarise(sphlan=n())

temp <- distances_filt_thresh %>% 
  group_by(relationship, strain) %>%
  summarise(count=n())

sphlan_species_tran_counts$tracm <- temp$count[match(paste(sphlan_species_tran_counts$relationship,
                                                                 sphlan_species_tran_counts$strain),
                                                           paste(temp$relationship, temp$strain))]

pdf <- sphlan_species_tran_counts %>% 
  filter(sphlan < tracm) %>%
  pivot_longer(cols = c('sphlan', 'tracm'), 
               names_to = "method", 
               values_to = "value")


ggplot(pdf %>% filter(relationship=="Same person (Infant)"), aes(x=strain, y=value, fill=method))+
  geom_col(position = 'dodge')+
  facet_wrap(~relationship, scales = 'free_x') +
  coord_flip() +
  # scale_y_sqrt() +
  scale_fill_brewer(palette = 6, type = 'qual') +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank()) +
  xlab("Species") +
  ylab("Number of inferred transmissions")

# ggsave("./figures/bbiome_sphlan_vs_tracm_species_counts.png", width = 12, height = 10)
# ggsave("./figures/bbiome_sphlan_vs_tracm_species_counts.pdf", width = 12, height = 10)


# Choose a host for further plotting/investigation

sphlan_breve <- sphlan_distances_filtered %>% 
  filter(strain=="Bifidobacterium breve") %>% 
  filter(relationship=="Same person (Infant)")
sphlan_breve$hostA <- distances_filt_thresh$subjectA[match(sphlan_breve$sampleA, gsub("_clean", "", distances_filt_thresh$sampleA))]

insphlan <- unique(sphlan_breve$pair)

# B. breve in Tracs but not StrainPhlan
tracm_only_breve <- distances_filt_thresh %>% 
  filter(strain=="Bifidobacterium breve") %>%
  filter(relationship=="Same person (Infant)") %>%
  filter(!gsub("_clean", "", pair) %in% insphlan)

obs_breve <- tracm_only_breve %>%
  group_by(subjectA) %>%
  summarise(
    count=n()
  ) %>%
  arrange(-count)

tb <- table(sphlan_breve$hostA)

obs_breve$sphlan_count <- tb[match(obs_breve$subjectA, names(tb))]

# B00560

interesting_pairs <- tracm_only_breve %>% filter(subjectA=='519128')

temp <- distances_filt_thresh %>% 
  filter(strain=="Bifidobacterium breve") %>%
  filter(relationship=="Same person (Infant)") %>%
  filter(subjectA=='B00560')

unique(c(temp$sampleA, temp$sampleB))

```

Plot allele frequency for example subject 'B00560' where StrainPhlAn did not detect some transmission of minority *Bifidobacterium breve* strains.

```{r}
allele_freqs_B00560 <- map_dfr(Sys.glob("./data/babybiome/21*_clean_ref_GCF_001025175.1_pileup.txt.gz"), function(f) {
  sample <- gsub("_clean.*", "", gsub(".*babybiome/", "", f))
  pile <- fread(f, col.names = c("contig",'pos','ref','alt','count')) %>%
    filter(ref!=alt)
  pile$splitcounts <- str_split(pile$count, ":")
  pile$splitcounts <- map(pile$splitcounts, ~ { colSums(matrix(as.numeric(str_split(.x[-1], ',', simplify = TRUE)), 
                                                               nrow=length(.x[-1])),na.rm = TRUE) })
  pile$allele <-  str_split(pile$alt, ",")
  pile$freq <- map(pile$splitcounts, ~ .x/sum(.x))
  
  pile <- pile[, .(counts = unlist(splitcounts),
                   allele = unlist(allele),
                   freq = unlist(freq)), by = .(contig, pos)] %>% 
    as_tibble() %>%
    add_column(sample=sample, .before=1)
  return(pile)
})

allele_freqs_B00560$time_point <- meta$Time_point[match(allele_freqs_B00560$sample, gsub("_clean", "", meta$Lane_ID))]

allele_freqs_B00560 <- allele_freqs_B00560 %>% filter(str_length(allele)<=1)

allele_freqs_B00560$time <- factor(paste('Day', allele_freqs_B00560$time_point), levels = c('Day 4', 'Day 7', 'Day 21'))

ggplot(allele_freqs_B00560 %>% 
         filter(pos>1.007e6, pos<1.0075e6, freq>0.01, freq<0.99),
       aes(x=pos, y=freq, col=allele)) +
  geom_point() +
  ggh4x::facet_wrap2(~time, axes = "all", remove_labels = "all", ncol = 1) +
  scale_color_brewer(type = 'qual', palette = 6) +
  scale_alpha_manual(values = c(0.4,1)) +
  scale_y_continuous(limits = c(0,1)) +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(),
        axis.text.x = element_text(angle = 45, hjust=1),
        legend.background = element_blank(), legend.title = element_blank()) +
  xlab("Position in reference") +
  ylab("Allele frequency") +
  guides(alpha='none')

# ggsave("./figures/bbiome_tracm_breve_example_zoom_B00560.png", width = 12, height = 10)
# ggsave("./figures/bbiome_tracm_breve_example_zoom_B00560.pdf", width = 12, height = 10)

pdf <- allele_freqs_B00560 %>% 
  filter(freq>0.01, freq<0.99) %>%
  group_by(allele, pos) %>%
  filter(n() >= 3) %>%
  ungroup()

pdf <- pdf %>%
  filter(paste(allele, pos) %in% sample(unique(paste(pdf$allele, pdf$pos)), 1000))

# Not using this at the moment
first_strain <- pdf %>% filter(freq>0.5) %>% filter(time=='Day 4')
pdf$strain <- ifelse(paste(pdf$pos, pdf$allele) %in% paste(first_strain$pos, first_strain$allele),
                     'Strain A', 'Strain B')

ggplot(pdf, aes(x=time, y=freq, col=strain, group=paste(pos, allele))) +
  scale_colour_manual(values = c('#762a83','#1b7837')) +
  ggrastr::rasterize(geom_line(alpha=0.05)) +
  theme_clean(base_size = 16) +
  theme(plot.background = element_blank(), legend.position = 'none') +
  xlab("Time point") +
  ylab("Allele frequency") +
  guides(alpha='none')


# ggsave("./figures/bbiome_tracm_breve_example_B00560.png", width = 12, height = 10)
# ggsave("./figures/bbiome_tracm_breve_example_B00560.pdf", width = 12, height = 10)
```
