---
title: "Streptococcus pneumoniae deep sequencing analysis"
output: html_document
date: '2023-05-10'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This Rmarkdown describes the TRACS analysis of the deep *Streptococcus pneumoniae* sequencing data originally published in Tonkin-Hill et al., 2022

> Tonkin-Hill, G. et al. Pneumococcal within-host diversity during colonization, transmission and treatment. Nat Microbiol 7, 1791â€“1804 (2022).

Consistent with that work a transmission generation time of 2 months and a clock rate of 5.3 SNPs/genome/year was used as input to the TransCluster algorithm


## Running TRACS

The TRACS align was run on each pair of fastq file using the following command using the same reference database as in the simulation analysis. This includes representative genomes from each major Global Pneumococcal Sequencing Cluster described in Gladstone et al., *EBioMedicine* (2019).

The following Bash commands accurately describe the execution process for running TRACS. However, they are not configured to run directly from this Rmarkdown document. Instead, these commands were executed separately on a server.

```
tracs align -i example_sample_1.fastq.gz example_sample_2.fastq.gz -p out_dir -o out_dir -t 20 --database gpsDB.zip --keep-all
```

The results of each alignment were then combined 

```
for dir in *; do
    if ls "${dir}"/*_sourmash_hits.csv 1> /dev/null 2>&1; then
        echo "${dir}" >> folder_list.txt
    fi
done
tracs combine -i folder_list.txt -o combined_alignments -t 30
```

SNP distances were then calculated as

```
for f in *_combined.fasta.gz
do
prefix=$(basename $f _combined.fasta.gz)
tracs dist --msa $f -o ${f}_distance_all.csv --snp_threshold 500000 --filter -t 10 -K 10000 --trans_rate 6 --clock_rate 5.3 --meta maela_sample_dates.csv
done
```

## Load results and generate plots

```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(data.table)
library(ape)
library(geosphere)
library(ggbeeswarm)
library(ggthemes)
library(splines)
library(statmod)
```


Load distance files

```{r}
files <- Sys.glob("./data/pneumo_deep/*distance_all.csv.gz")

distances <- map_dfr(files, ~{
  ref <- gsub(".*/", "", gsub("\\.velvet.*", "", .x))
  gpsc <- gsub("_.*", "", ref)
  
  d <- fread(.x) %>%
    as_tibble()
  
  if (nrow(d) < 1) return(NULL)
  
  return(d %>%
    add_column(gpsc=gpsc) %>%
    add_column(ref=ref))
  
})

meta <- fread("./data/pneumo_deep/combined_pnc_swab_house_trans_rf.tsv.gz") %>%
  as_tibble()
manifest <- fread("./data/pneumo_deep/manifest_w_QC.csv.gz") %>%
  as_tibble()
```

Add metadata

```{r}
distances$specA <- manifest$`DONOR ID (required for EGA)`[match(distances$sampleA, manifest$`Lane Name`)]
distances$specB <- manifest$`DONOR ID (required for EGA)`[match(distances$sampleB, manifest$`Lane Name`)]

distances_filt <- distances[(distances$specA %in% meta$specnum) & (distances$specB %in% meta$specnum),]

distances_filt$geo_dist <- map2_dbl(distances_filt$specA, distances_filt$specB, ~{
  indexA <- which(meta$specnum==.x)[[1]]
  indexB <- which(meta$specnum==.y)[[1]]
  
  return(distHaversine(c(meta$house_longitude[indexA],meta$house_latitude[indexA]),
                 c(meta$house_longitude[indexB],meta$house_latitude[indexB])))

})

distances_filt$houseA <- meta$codenum.x[match(distances_filt$specA, meta$specnum)]
distances_filt$houseB <- meta$codenum.x[match(distances_filt$specB, meta$specnum)]

distances_filt$hostA <- paste(distances_filt$houseA,
                              meta$category[match(distances_filt$specA, meta$specnum)])
distances_filt$hostB <- paste(distances_filt$houseB,
                              meta$category[match(distances_filt$specB, meta$specnum)])

```

We now generate plots of geographic distance versus the expected number of intermediate hosts separating two genomes, calculated using the extended TransCluster algorithm in TRACS.

With a clock rate of 5.3 SNPs per genome per year, genome pairs separated by more than 150 SNPs are unlikely to share a most recent common ancestor within the camp, established in 1984; therefore, these pairs are excluded from the plot. Additionally, since there are very few household pairs separated by more than 2 km within the camp, these distant pairs are also excluded.

```{r}
distances_filt_single <- distances_filt %>% 
  arrange(`SNP distance`) %>%
  filter(!is.na(geo_dist))
distances_filt_single <- distances_filt_single[!duplicated(paste(distances_filt_single$sampleA, distances_filt_single$sampleB)),]
distances_filt_single <- distances_filt_single[!duplicated(paste(distances_filt_single$sampleB, distances_filt_single$sampleA)),]

distances_filt_single$within_family <- distances_filt_single$houseA==distances_filt_single$houseB

distances_filt_single <- distances_filt_single[distances_filt_single$hostA!=distances_filt_single$hostB,]

distances_filt_single <- distances_filt_single %>% filter(!within_family)


min_distances_filt_single <- distances_filt_single %>% 
  arrange(`expected K`) %>%
  group_by(hostA, hostB) %>%
  slice_head(n = 1)


ggplot(min_distances_filt_single %>% 
         filter(geo_dist<2000) %>%
         filter(`filtered SNP distance`<150),
       aes(x=geo_dist, y=`expected K`)) +
  geom_smooth(method='loess', se=T) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank()) +
  xlab("Geographic distance (meters)") +
  ylab("Expected number of intermediate hosts")

# ggsave("./figures/pneumo_deep_Ek_vs_geog_smooth.png", width = 10, height = 7)
# ggsave("./figures/pneumo_deep_Ek_vs_geog_smooth.pdf", width = 10, height = 7)

gpsc_count <- sort(table(min_distances_filt_single$gpsc), decreasing = TRUE)
keep <- names(gpsc_count)[1:3]

min_distances_filt_single <- distances_filt_single %>%
  arrange(`expected K`) %>%
  group_by(hostA, hostB, gpsc) %>%
  slice_head(n = 1) 


ggplot(min_distances_filt_single %>% 
         filter(geo_dist<2000) %>% 
         filter(`filtered SNP distance`<150) %>%
         filter(gpsc %in% keep), 
       aes(x=geo_dist, y=`expected K`, colour=gpsc, fill=gpsc)) +
  geom_smooth(method='loess', se=T) +
  facet_wrap(~gpsc) +
  scale_color_brewer(type = 'q', palette = 1) +
  scale_fill_brewer(type = 'q', palette = 1) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank(), legend.background = element_blank()) +
  xlab("Geographic distance (meters)") +
  ylab("Expected number of intermediate hosts")

# ggsave("./figures/pneumo_deep_Ek_vs_geog_smooth_top3_GPSCs.png", width = 12, height = 7)
# ggsave("./figures/pneumo_deep_Ek_vs_geog_smooth_top3_GPSCs.pdf", width = 12, height = 7)
```