---
title: "SARS-CoV-2 deep sequencing analysis"
output: html_document
date: '2023-05-10'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This Rmarkdown describes the TRACS analysis of the deep SARS-CoV-2 sequencing data originally published in Tonkin-Hill et al., 2021

Consistent with that work a transmission generation time of 5 days and a clock rate of 29.903 SNPs/genome/year was used as input to the expanded TransCluster algorithm implemented in TRACS

## Running TRACS

The TRACS align was run on each pair of fastq files using the following command using the MN908947.3 SARS-CoV-2 reference genome.

The following Bash commands accurately describe the execution process for running TRACS. However, they are not configured to run directly from this Rmarkdown document. Instead, these commands were executed separately on a server.

Initially TRACS was run in the consensus mode where only the dominant allele at each position within the alignment is considered

```
tracs align -i example_sample_1.fastq.gz example_sample_2.fastq.gz -p out_dir -o out_dir_consensus  -t 15 --refseqs nCoV-2019.reference.fasta --consensus --either-strand
```

TRACS was then run in the standard mode which considers minority variants

```
tracs align -i example_sample_1.fastq.gz example_sample_2.fastq.gz -p -p out_dir -o out_dir_default"  -t 15 --refseqs nCoV-2019.reference.fasta --either-strand
```

The TRACS distance command was then run on the resulting alignments generated by each method. The `sc2_tonkinhill_meta` data contains the sample dates.

```
tracs distance --msa filt_nCoV-2019_consensus.reference_combined.fasta -o sc2_distance_consensus.csv --snp_threshold 100 -t 10 -K 1000 --trans_rate 73 --clock_rate 29.903 --meta sc2_tonkinhill_meta.csv

tracs distance --msa filt_nCoV-2019_default.reference_combined.fasta -o sc2_distance_default.csv --snp_threshold 100 -t 10 -K 1000 --trans_rate 73 --clock_rate 29.903 --meta sc2_tonkinhill_meta.csv
```

## Load results and generate plots

```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(data.table)
library(ggthemes)
library(ape)
library(ggraph)
library(igraph)
library(ggnetwork)
```

Here we load the original MSA using the SARS-CoV-2 specific pipeline as was run in Tonkin-Hill et al., 2021.

```{r}
dist_orig <- fread("./data/sc2/MA_combined_consensus_replicates_dists.csv.gz") %>%
  as_tibble() %>%
  filter(V1!="MN908947.3") %>%
  filter(grepl(".*run1", V2)) %>%
  filter(grepl(".*run1", V1))
dist_orig$V1 <- gsub("_run1", "", dist_orig$V1)
dist_orig$V2 <- gsub("_run1", "", dist_orig$V2)
dist_orig$pair <- map2_chr(dist_orig$V1, dist_orig$V2, ~ paste(sort(c(.x,.y)), collapse = '-'))
colnames(dist_orig)[[3]] <- "SNP distance"


dist_consensus <- fread("./data/sc2/sc2_distance_consensus.csv.gz") %>%
  as_tibble()
dist_consensus$pair <- map2_chr(dist_consensus$sampleA, dist_consensus$sampleB, ~ paste(sort(c(.x,.y)), collapse = '-'))
```

Load TRACS distances

```{r}
distances <- fread("./data/sc2/sc2_distance_default.csv.gz") %>% as_tibble()
distances$pair <-  map2_chr(distances$sampleA, distances$sampleB, ~ paste(sort(c(.x,.y)), collapse = '-'))

meta <- fread("./data/sc2/sc2_tonkinhill_meta.csv.gz") %>% as_tibble()
```

Merge and add metadata

```{r}
distances$sampleA_alt <- meta$sample_id[match(distances$sampleA, meta$Accession)]
distances$sampleB_alt <- meta$sample_id[match(distances$sampleB, meta$Accession)]
distances$pair_alt <-  map2_chr(distances$sampleA_alt, distances$sampleB_alt, ~ paste(sort(c(.x,.y)), collapse = '-'))
distances$pair <- map2_chr(distances$sampleA_alt, distances$sampleB_alt, ~ paste(sort(c(.x,.y)), collapse = '-'))
distances$pair_accession <- map2_chr(distances$sampleA, distances$sampleB, ~ paste(sort(c(.x,.y)), collapse = '-'))
distances$snp_orig <- dist_orig$`SNP distance`[match(distances$pair, dist_orig$pair)]
distances$snp_consensus <- dist_consensus$`SNP distance`[match(distances$pair_accession, dist_consensus$pair)]
```

Compare distances estimated by each approach across all samples

```{r}
gg1 <- ggplot(distances %>% filter(`sites considered`>0), aes(x=snp_orig, y=`SNP distance`)) +
  geom_bin2d(binwidth=1) +
  geom_abline(slope = 1, intercept = 0, col='red') +
  scale_fill_distiller(type = 'seq', palette = 3, values = c(0,0.1,0.25,0.5,0.75,0.9,1), direction = 1) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank()) +
  xlab("Original SNP distance") +
  ylab("TRACS distance")

gg1

# ggsave("./figures/sc2_tracm_vs_orig.png", width = 12, height = 7)
# ggsave("./figures/sc2_tracm_vs_orig.pdf", width = 12, height = 7)

gg2 <- ggplot(distances %>% filter(`sites considered`>0), aes(x=snp_consensus, y=`SNP distance`)) +
  geom_bin2d(binwidth=1) +
  geom_abline(slope = 1, intercept = 0, col='red') +
  scale_fill_distiller(type = 'seq', palette = 3, values = c(0,0.1,0.25,0.5,0.75,0.9,1), direction = 1) +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank()) +
  xlab("Consensus SNP distance") +
  ylab("TRACS distance")

gg2

# ggsave("./figures/sc2_tracm_vs_consensus.png", width = 12, height = 7)
# ggsave("./figures/sc2_tracm_vs_consensus.pdf", width = 12, height = 7)
```

## Investigate known mixtures

We now investigate the known mixtures originally identified in Tonkin-Hill et al., 2021.

```{r}
mixture_tests <- fread("./data/sc2/mixture_tests_revised.csv.gz", data.table = FALSE) %>% as_tibble()
mixture_tests_mm <- mixture_tests %>% 
  filter(n_mismatch<2) %>%
  filter(n_extra_explained>1) %>%
  filter(estimate>0) %>%
  arrange(p.value)
mixture_tests_mm <- mixture_tests_mm[!duplicated(mixture_tests_mm$sampleA),]

# remove those that on manual inspection don't appear to be convincingly mixtures
mixture_tests_mm <- mixture_tests_mm[!mixture_tests_mm$sampleA %in% c(
  'CAMB-7A326','CAMB-75E93', 'CAMB-71F76', 'CAMB-71D30',
  'CAMB-7657F', 'CAMB-776F3', 'CAMB-76B31', 'CAMB-76BC8',
  'CAMB-7A71B', 'CAMB-7780C', 'CAMB-72494', 'CAMB-79433',
  'CAMB-77CB5', 'CAMB-775E7', 'CAMB-75FFA', 'CAMB-7A890',
  'CAMB-76B9B', 'CAMB-7674C', 'CAMB-77FBC'),]

mixture_tests_mm$pair <- map2_chr(mixture_tests_mm$sampleA, mixture_tests_mm$sampleB, ~ paste(sort(c(.x,.y)), collapse = '-'))
```

Create plot

```{r}
mix_dists <- distances %>% filter(pair_alt %in% mixture_tests_mm$pair)

pdf <- tibble(method=rep(c("Consensus", "Original", "Trac'm"), each=nrow(mix_dists)),
       difference=c(mix_dists$snp_consensus,
                    mix_dists$snp_orig,
                    mix_dists$`SNP distance`))

ggplot(pdf, aes(x=method, y=difference, colour=method)) +
  geom_boxplot(outlier.color = NA) +
  ggbeeswarm::geom_beeswarm(size=3) +
  ggsci::scale_color_aaas() +
  theme_clean(base_size = 20) +
  theme(plot.background = element_blank(),
        legend.background = element_blank()) +
  xlab("") +
  ylab("SNP distance")

# ggsave("./figures/sc2_tracm_vs_classic_mixed_infection_boxplot.png", width = 12, height = 7)
# ggsave("./figures/sc2_tracm_vs_classic_mixed_infection_boxplot.pdf", width = 12, height = 7)
```

Generate network plot

```{r}
mixture_tests_mm_strict <- mixture_tests_mm

keep <- unique(c(mixture_tests_mm_strict$sampleA, mixture_tests_mm_strict$sampleB)) 
keep <- keep[keep %in% meta$sample_id[!meta$adm2 %in% c("OTHER")]]

# We use a 0 SNP threshold
snp_threshold <- 0

mix_nei_dists <- distances %>% 
  filter(sampleA_alt %in% keep, sampleB_alt %in% keep) %>%
  filter(`SNP distance` <= snp_threshold)

mix_nei_dists$method <- ifelse(mix_nei_dists$snp_orig > snp_threshold, "Tracm", "Consensus")

keep <- unique(c(mix_nei_dists$sampleA[mix_nei_dists$method=="Consensus"],
                 mix_nei_dists$sampleB[mix_nei_dists$method=="Consensus"]))

mix_nei_dists <- mix_nei_dists %>%
  filter((sampleA %in% keep) & (sampleB %in% keep))

edges <- mix_nei_dists[, c('sampleA','sampleB', 'expected K','SNP distance','method')]

# Create an igraph object from the tibble
graph <- graph_from_data_frame(edges, directed = FALSE)

graph <- set_vertex_attr(graph, "node_type", value = meta$adm2[match(igraph::vertex_attr(graph)$name, meta$Accession)])

# Generate the network plot
ggraph(graph, layout = "auto") + 
    geom_edge_fan(aes(linetype = method), width = 1) +
    geom_node_point(aes(color = node_type), size = 5) +
   	ggsci::scale_color_npg() +
    theme_void() +
    theme(plot.background = element_blank()) +
    ggtitle("Transmission Network")

# ggsave("./figures/sc2_tracm_vs_classic_network_graph.png", width = 12, height = 7)
# ggsave("./figures/sc2_tracm_vs_classic_network_graph.pdf", width = 12, height = 7)
```


